### PR TITLE
Closes #542 - make clustering possible by adding locking

### LINKED ISSUE #542: make adapter cluster capable
Enable the adapter to run on a cluster, i.e. allow multiple instances of the adapter to run concurrently

### PR DESCRIPTION
**Pull request by ryzheboka**
Release Notes:
```
Add the possiblility to run multiple instances of Adapter connected to the same TASKANA simultaneously. This is accomplished by locking the database entries that are currently worked on by an instance. This way, it's impossible for multiple instances to work on the same entry simultaneously. Locking of the entries in the events table is configured by the new parameter taskana.adapter.events.lockDuration that describes the duration of the lock for one event in seconds. The parameter is set to zero by default. To enable running multiple Adapter instances, the parameter needs to be configured to a suitable positive number. A good value for taskana.adapter.events.lockDuration depends on the application.
```
Updated documentation:
https://taskana.atlassian.net/wiki/spaces/TAS/pages/996966415/Adapter+Properties
https://taskana.atlassian.net/wiki/spaces/TAS/pages/1649377329/Failed+task+creation+handling+configuration+monitoring

---
 
### For the submitter:
- [x] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- [ ] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [ ] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
- [ ] I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I put the ticket in review

### Verified by the reviewer:
- [ ] Commit message format → ADPT-XXX: Your commit message.
- [ ] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [ ] SonarCloud analysis meets our standards
- [ ] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [ ] PR fulfills the ticket
- [ ] Edge cases and unwanted side effects are tested
- [ ] Readability

### CODE REVIEW

File: taskana-adapter-camunda-outbox-rest/src/main/resources/taskana-outbox.properties [hunk 1]
```diff
@@ -7,12 +7,12 @@ taskana.adapter.outbox.duration.between.task.creation.retries = PT1H
 #taskana.adapter.outbox.datasource.jndi=java:jboss/datasources/ProcessEnginePostgres
 #taskana.adapter.outbox.datasource.jndi=jdbc/ProcessEngine
 
-#taskana.adapter.outbox.datasource.driver=org.postgresql.Driver
-#taskana.adapter.outbox.datasource.url=jdbc:postgresql://localhost:5432/postgres
-#taskana.adapter.outbox.datasource.username=postgres
-#taskana.adapter.outbox.datasource.password=postgres
+taskana.adapter.outbox.datasource.driver=org.postgresql.Driver
```
Comment (gitgoodjhe) [inline taskana-adapter-camunda-outbox-rest/src/main/resources/taskana-outbox.properties:10]: do we need these changes?

File: taskana-adapter-camunda-spring-boot-example/src/main/resources/application.properties [hunk 2]
```diff
@@ -11,8 +11,9 @@ logging.level.pro.taskana=DEBUG
 #logging.level.pro.taskana=info
 
 ## Set Server Port for Adapter
-server.port = 8082
+server.port = 8083
```
Comment (gitgoodjhe) [inline taskana-adapter-camunda-spring-boot-example/src/main/resources/application.properties:14]: do we need these changes in this file?

File: taskana-adapter-camunda-spring-boot-test/src/main/resources/application.properties [hunk 3]
```diff
@@ -26,11 +26,11 @@ logging.level.pro.taskana=INFO
 taskana.adapter.run-as.user=peter
 taskana.adapter.camunda.claiming.enabled=true
 taskana.adapter.scheduler.run.interval.for.start.taskana.tasks.in.milliseconds=1000
-taskana.adapter.scheduler.run.interval.for.complete.referenced.tasks.in.milliseconds=1000
-taskana.adapter.scheduler.run.interval.for.claim.referenced.tasks.in.milliseconds=1000
-taskana.adapter.scheduler.run.interval.for.cancel.claim.referenced.tasks.in.milliseconds=1000
-taskana.adapter.scheduler.run.interval.for.check.finished.referenced.tasks.in.milliseconds=1000
-taskana.adapter.scheduler.run.interval.for.retries.and.blocking.taskevents.in.milliseconds = 10000
+taskana.adapter.scheduler.run.interval.for.complete.referenced.tasks.in.milliseconds=1000000000
```
Comment (gitgoodjhe) [inline taskana-adapter-camunda-spring-boot-test/src/main/resources/application.properties:29]: please revert this

File: taskana-adapter-camunda-spring-boot-test/src/main/resources/application.properties [hunk 4]
```diff
@@ -79,10 +80,10 @@ taskana.adapter.mapping.default.objectreference.value=DEFAULT_VALUE
 #H2
 #camunda.datasource.jdbcUrl= jdbc:h2:./camunda-db;NON_KEYWORDS=KEY,VALUE;DB_CLOSE_DELAY=-1;MVCC=TRUE;DB_CLOSE_ON_EXIT=FALSE
 #camunda.datasource.jdbcUrl=jdbc:h2:mem:camunda;NON_KEYWORDS=KEY,VALUE;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE
- camunda.datasource.jdbcUrl=jdbc:h2:mem:camunda;NON_KEYWORDS=KEY,VALUE;IGNORECASE=TRUE;LOCK_MODE=0;DB_CLOSE_ON_EXIT=FALSE;DB_CLOSE_DELAY=-1
- camunda.datasource.driverClassName=org.h2.Driver
- camunda.datasource.username=sa
- camunda.datasource.password=sa
+ #camunda.datasource.jdbcUrl=jdbc:h2:mem:camunda;NON_KEYWORDS=KEY,VALUE;IGNORECASE=TRUE;LOCK_MODE=0;DB_CLOSE_ON_EXIT=FALSE;DB_CLOSE_DELAY=-1
+ #camunda.datasource.driverClassName=org.h2.Driver
```
Comment (gitgoodjhe) [inline taskana-adapter-camunda-spring-boot-test/src/main/resources/application.properties:84]: do we need these changes?

**Review by gitgoodjhe — CHANGES_REQUESTED** [review]

File: camunda-outbox-example-boot/src/main/resources/application.properties [hunk 5]
```diff
@@ -1,14 +1,16 @@
-server.port=8081
+server.port=8085
```
[the remark this answers is not in the document]
Comment (ryzheboka) [reply camunda-outbox-example-boot/src/main/resources/application.properties:1]: No, I will change it back :)

File: taskana-adapter-camunda-outbox-rest/src/main/resources/taskana-outbox.properties (hunk 1 above)
Comment (ryzheboka) [reply taskana-adapter-camunda-outbox-rest/src/main/resources/taskana-outbox.properties:10]: We need to change the port, the url and the password to comply to those in our postgres docker container.

File: camunda-outbox-example-boot/src/main/resources/application.properties [hunk 6]
```diff
@@ -1,14 +1,16 @@
...
+
 
 server.servlet.context-path=/example-context-root
 camunda.bpm.admin-user.id=demo
 camunda.bpm.admin-user.first-name=demo
 camunda.bpm.admin-user.password=demo
 camunda.bpm.admin-user.last-name=demo
 
-camunda.bpm.database.type=h2
+camunda.bpm.database.type=postgres
```
[the remark this answers is not in the document]
Comment (ryzheboka) [reply camunda-outbox-example-boot/src/main/resources/application.properties:13]: No, I will change it back :)

File: taskana-adapter-camunda-spring-boot-example/src/main/resources/application.properties (hunk 2 above)
Comment (ryzheboka) [reply taskana-adapter-camunda-spring-boot-example/src/main/resources/application.properties:14]: No, I will change it back :)

File: taskana-adapter-camunda-spring-boot-test/src/main/resources/application.properties (hunk 4 above)
Comment (ryzheboka) [reply taskana-adapter-camunda-spring-boot-test/src/main/resources/application.properties:84]: No, I will change it back

File: taskana-adapter-camunda-spring-boot-test/src/main/resources/taskana-outbox.properties [hunk 7]
```diff
@@ -5,15 +5,15 @@ taskana.adapter.outbox.initial.number.of.task.creation.retries = 3
 taskana.adapter.outbox.duration.between.task.creation.retries = PT1S
 
 
-#taskana.adapter.outbox.datasource.driver=org.postgresql.Driver
-#taskana.adapter.outbox.datasource.url=jdbc:postgresql://localhost:5432/postgres
-#taskana.adapter.outbox.datasource.username=postgres
-#taskana.adapter.outbox.datasource.password=postgres
+taskana.adapter.outbox.datasource.driver=org.postgresql.Driver
+taskana.adapter.outbox.datasource.url=jdbc:postgresql://localhost:5102/postgres
```
[the remark this answers is not in the document]
Comment (ryzheboka) [reply taskana-adapter-camunda-spring-boot-test/src/main/resources/taskana-outbox.properties:9]: Changing the port to coply with our existing Docker container is good. I will comment it out, so our "default" configured database is h2

File: taskana-adapter-camunda-outbox-rest/src/main/java/pro/taskana/adapter/camunda/outbox/rest/service/CamundaTaskEventsService.java [hunk 8]
```diff
@@ -156,34 +187,51 @@ public void decreaseRemainingRetriesAndLogError(String eventIdAndErrorLog) {
...
+      ;
+      lockEvents(ids, lockDuration, connection);
 
     } catch (Exception e) {
       LOGGER.warn("Caught Exception while trying to retrieve failed events from the outbox", e);
+    } finally {
+      try {
+        assert connection != null;
+        assert ids != null;
+        unlockEvents(ids, connection);
```
Comment (gitgoodjhe) [inline taskana-adapter-camunda-outbox-rest/src/main/java/pro/taskana/adapter/camunda/outbox/rest/service/CamundaTaskEventsService.java:230]: please log the ids of the unlocked Events

**Review by gitgoodjhe — CHANGES_REQUESTED** [review]

File: taskana-adapter-camunda-outbox-rest/src/main/java/pro/taskana/adapter/camunda/outbox/rest/service/CamundaTaskEventsService.java [hunk 9]
```diff
@@ -330,24 +388,59 @@ public CamundaTaskEvent getEvent(int id) throws CamundaTaskEventNotFoundExceptio
...
+          }
+        }
+      }
     } catch (Exception e) {
       LOGGER.warn("Caught Exception while trying to retrieve all events from the outbox", e);
     }
     return camundaTaskEvents;
   }
 
+  public void unlockEventsForId(Integer eventId) {
```
Comment (gitgoodjhe) [inline taskana-adapter-camunda-outbox-rest/src/main/java/pro/taskana/adapter/camunda/outbox/rest/service/CamundaTaskEventsService.java:436]: this always unlocks exactely one event, right? If yes, what do you think about naming it unlockEventForId?

File: taskana-adapter-camunda-outbox-rest/src/main/java/pro/taskana/adapter/camunda/outbox/rest/controller/CamundaTaskEventsController.java [hunk 10]
```diff
@@ -95,6 +95,18 @@ public Response decreaseRemainingRetriesAndLogError(
     return Response.status(204).build();
   }
 
+  @Path(Mapping.URL_UNLOCK_EVENT)
+  @POST
+  @Consumes(MediaType.APPLICATION_JSON)
+  public Response unlockEvents(
```
Comment (gitgoodjhe) [inline taskana-adapter-camunda-outbox-rest/src/main/java/pro/taskana/adapter/camunda/outbox/rest/controller/CamundaTaskEventsController.java:101]: this always unlocks exactely one event, right? If yes, what do you think about naming it unlockEvent?

File: taskana-adapter-camunda-spring-boot-example/src/main/resources/application.properties [hunk 11]
```diff
@@ -45,13 +45,16 @@ taskana.datasource.driverClassName = org.h2.Driver
 taskana.datasource.username = sa
 taskana.datasource.password = sa
 taskana.schemaName=TASKANA
+taskana.lockDuration=300
```
Comment (gitgoodjhe) [inline taskana-adapter-camunda-spring-boot-example/src/main/resources/application.properties:48]: This is the lock duration for the events, right? Should we maybe name it taskana.adapter.events.lockDuration? To be a little more consistent with e.g. taskana.adapter.scheduler.run.interval etc.

File: taskana-adapter-camunda-spring-boot-example/src/main/resources/application.properties [hunk 12]
```diff
@@ -45,13 +45,16 @@ taskana.datasource.driverClassName = org.h2.Driver
...
+taskana.adapter.fromTaskanaToAdapterBatchSize=1
```
Comment (gitgoodjhe) [inline taskana-adapter-camunda-spring-boot-example/src/main/resources/application.properties:49]: Alternatives: taskana.adapter.taskana.sync.batchSize or taskana.adapter.sync.taskana.batchSize

**Review by gitgoodjhe — CHANGES_REQUESTED** [review]

File: taskana-adapter-camunda-spring-boot-example/src/main/resources/application.properties (hunk 11 above)
Comment (ryzheboka) [reply taskana-adapter-camunda-spring-boot-example/src/main/resources/application.properties:48]: Yes, sounds great!

File: taskana-adapter-camunda-outbox-rest/src/main/java/pro/taskana/adapter/camunda/outbox/rest/controller/CamundaTaskEventsController.java (hunk 10 above)
Comment (ryzheboka) [reply taskana-adapter-camunda-outbox-rest/src/main/java/pro/taskana/adapter/camunda/outbox/rest/controller/CamundaTaskEventsController.java:101]: Yes, I will rename it

File: taskana-adapter-camunda-outbox-rest/src/main/java/pro/taskana/adapter/camunda/outbox/rest/service/CamundaTaskEventsService.java (hunk 9 above)
Comment (ryzheboka) [reply taskana-adapter-camunda-outbox-rest/src/main/java/pro/taskana/adapter/camunda/outbox/rest/service/CamundaTaskEventsService.java:436]: Yes, I will rename it

File: taskana-adapter-camunda-spring-boot-example/src/main/resources/application.properties (hunk 12 above)
Comment (ryzheboka) [reply taskana-adapter-camunda-spring-boot-example/src/main/resources/application.properties:49]: taskana.adapter.sync.taskana.batchSize sounds nice

**Pull request merged**